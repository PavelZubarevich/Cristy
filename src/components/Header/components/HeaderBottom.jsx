import { IoCartOutline, IoSearchOutline } from 'react-icons/io5';
import { Link } from 'react-router-dom';

function HeaderBottom() {
  return (
    <div className="header__bottom container">
      <div className="burger">
        <span />
      </div>
      <h1 className="header__logo"><Link to="/">Cristy</Link></h1>
      <div className="header__buttons">
        <IoCartOutline className="header__cart" />
        <IoSearchOutline className="header__search" />
      </div>
    </div>
  );
}

export default HeaderBottom;
