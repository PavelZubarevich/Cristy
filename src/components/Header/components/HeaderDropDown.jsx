import PropTypes from 'prop-types';
import { useState, useEffect, useRef } from 'react';

function HeaderDropDown({ title, children }) {
  const [show, setShow] = useState(false);
  const ref = useRef();

  const handleButtonClick = () => {
    setShow((prev) => !prev);
  };

  const closeDropDown = (e) => {
    if (e.target.closest('.headerDropDown') !== ref.current)setShow(false);
  };

  useEffect(() => {
    document.addEventListener('click', closeDropDown);

    return () => {
      document.removeEventListener('click', closeDropDown);
    };
  }, []);

  return (
    <div className="headerDropDown" ref={ref}>
      <button type="button" className={`headerDropDown__title ${show ? 'active' : ''}`} onClick={handleButtonClick} aria-haspopup="true" aria-expanded={show}>{title}</button>
      {show && <div className="headerDropDown__content">{children}</div>}
    </div>
  );
}

export default HeaderDropDown;

HeaderDropDown.propTypes = {
  title: PropTypes.string,
  children: PropTypes.node.isRequired,
};

HeaderDropDown.defaultProps = {
  title: 'Default title',
};
