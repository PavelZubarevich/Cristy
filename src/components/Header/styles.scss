@import '../../styles/vars.scss';

.header {

  background-color: #fff;

  &__top {
    border-bottom: 1px solid $secondaryText;
  }

  &__topContent {
    display: flex;
    justify-content: space-between;
    align-items: center;
    height: 40px;
  }

  &__topLabel {
    font-family: 'Lora';
    font-weight: 400;
    font-size: 13px;
    color: $secondaryText;
  }

  &__topMenu {
    display: flex;
    height: 100%;

    &>li {
      border-right: 1px solid $secondaryText;
    }
  }

  &__langList {
    &>li {
      margin-bottom: 10px;

      &:last-child {
        margin-bottom: 0;
      }
    }
  }

  &__login {
    margin-bottom: 20px;
  }

  &__createAccount {
    font-family: 'Lora';
    font-weight: 400;
    font-size: 13px;
    color: $primaryText;
    transition: 0.3s;

    &:hover {
      color: #000;
    }
  }

  &__bottom {
    height: 120px;
    display: flex;
    justify-content: space-between;
    align-items: center;
  }

  &__logo {
    font-family: 'Lora';
    font-weight: 700;
    font-style: italic;
    font-size: 36px;
    color: $primaryDarkText;
  }

  &__search,
  &__cart {
    color: $primaryDarkText;
    cursor: pointer;
  }

  &__cart {
    font-size: 26px;

  }

  &__search {
    font-size: 24px;
    margin-left: 50px;
  }
}

.headerDropDown {
  height: 100%;
  position: relative;

  &__title {
    font-size: 13px;
    font-family: 'Lora';
    font-weight: 400;
    color: $secondaryText;
    text-transform: capitalize;
    background-color: transparent;
    border: none;
    cursor: pointer;
    padding: 0 35px 0 20px;
    height: 100%;
    position: relative;
    transition: 0.3s;

    &::after {
      content: '';
      display: block;
      width: 0px;
      height: 0px;
      border-style: solid;
      border-width: 5px 5px 5px 5px;
      border-color: $secondaryText transparent transparent transparent;
      position: absolute;
      right: 15px;
      top: calc(50% - 2px);
      transition: 0.3s;
    }

    &:hover {
      color: #000;

      &::after {
        border-color: #000 transparent transparent transparent;
      }
    }

    &.active {
      &::after {
        transform: rotateX(180deg) translateY(5px);
      }
    }
  }

  &__content {
    position: absolute;
    top: 41px;
    padding: 20px;
    background-color: #fff;
    box-shadow: 0px 1px 5px 0px rgba(153, 153, 153, 0.6);
    display: flex;
    flex-direction: column;
    align-items: center;
  }
}

.burger {
  width: 20px;
  height: 14px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  cursor: pointer;

  & span {
    width: 100%;
    height: 2px;
    background-color: $primaryDarkText;
  }

  &:before,
  &:after {
    content: '';
    display: block;
    width: 100%;
    height: 2px;
    background-color: $primaryDarkText;
  }
}