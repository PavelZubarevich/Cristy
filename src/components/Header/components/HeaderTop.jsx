import '../styles.scss';
import { useTranslation } from 'react-i18next';
import HeaderDropDown from './HeaderDropDown';
import Button from '../../common/Button';

function HeaderTop() {
  const { t, i18n } = useTranslation();

  const changeLang = (lang) => {
    i18n.changeLanguage(lang);
  };

  return (
    <div className="header__top">
      <div className="container">
        <div className="header__topContent">
          <p className="header__topLabel">{t('FreeShippingOverText', { value: 111 })}</p>
          <ul className="header__topMenu">
            {i18n.options.supportedLngs && i18n.options.supportedLngs.length > 1 && (
            <li>
              <HeaderDropDown title={i18n.language}>
                <ul className="header__langList">
                  {i18n.options.supportedLngs.map((elem) => {
                    if (elem !== 'cimode' && elem !== i18n.language) {
                      return (
                        <li key={elem}>
                          <Button text={elem} variant="text" color="primaryText" size="sm" onClick={() => changeLang(elem)} />
                        </li>
                      );
                    }
                    return false;
                  })}
                </ul>
              </HeaderDropDown>
            </li>
            )}
            <li>
              <HeaderDropDown title="USD">
                <div />
              </HeaderDropDown>
            </li>
            <li>
              <HeaderDropDown title={t('LoginLabel')}>
                <>
                  <Button text={t('LoginLabel')} variant="contained" className="header__login" />
                  <a href="/" className="header__createAccount">{t('CreateAccountText')}</a>
                </>
              </HeaderDropDown>
            </li>
          </ul>
        </div>
      </div>
    </div>
  );
}

export default HeaderTop;
