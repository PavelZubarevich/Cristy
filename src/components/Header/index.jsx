import HeaderBottom from './components/HeaderBottom';
import HeaderTop from './components/HeaderTop';
import './styles.scss';

function Header() {
  return (
    <header className="header">
      <HeaderTop />
      <HeaderBottom />
    </header>
  );
}

export default Header;
