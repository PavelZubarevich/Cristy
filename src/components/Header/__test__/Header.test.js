import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import Header from '..';
import HeaderDropDown from '../components/HeaderDropDown';
import HeaderTop from '../components/HeaderTop';
import HeaderBottom from '../components/HeaderBottom';

const mockChangeLanguage = jest.fn();

jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (str) => str,
    i18n: {
      changeLanguage: mockChangeLanguage,
      options: {
        supportedLngs: ['ru', 'en', 'cimode'],
      },
      language: 'ru',
    },
  }),
}));

describe('Header', () => {
  it('header is rendered', () => {
    render(<Header />, { wrapper: BrowserRouter });
  });
});

describe('Header top', () => {
  it('header top is rendered', () => {
    render(<HeaderTop />, { wrapper: BrowserRouter });
    expect(document.body).toMatchSnapshot();
  });

  it('header top is', () => {
    render(<HeaderTop />, { wrapper: BrowserRouter });
    userEvent.click(screen.getByText('ru'));
    userEvent.click(screen.getByText('en'));
    expect(mockChangeLanguage).toBeCalledWith('en');
  });
});

describe('Header bottom', () => {
  it('header is rendered', () => {
    render(<HeaderBottom />, { wrapper: BrowserRouter });
  });
});

describe('Header dropdown', () => {
  it('dropdown is renders', () => {
    render(<HeaderDropDown>test</HeaderDropDown>);
    expect(screen.getByRole('button')).toBeInTheDocument();
  });

  it('dropdown renders with default title', () => {
    render(<HeaderDropDown>test</HeaderDropDown>);
    expect(screen.getByText(/Default title/i)).toBeInTheDocument();
  });

  it('dropdown renders with title', () => {
    render(<HeaderDropDown title="my title">test</HeaderDropDown>);
    expect(screen.getByText(/my title/i)).toBeInTheDocument();
  });

  it('dropdown works', () => {
    render(<HeaderDropDown title="my title">test</HeaderDropDown>);
    userEvent.click(screen.getByRole('button'));
    expect(screen.getByText(/test/i)).toBeInTheDocument();
    userEvent.click(screen.getByRole('button'));
    expect(screen.queryByText(/test/i)).not.toBeInTheDocument();
  });

  it('dropdown closes on click outside', () => {
    render(<HeaderDropDown title="my title">test</HeaderDropDown>);
    userEvent.click(screen.getByRole('button'));
    expect(screen.getByText(/test/i)).toBeInTheDocument();
    userEvent.click(document.body);
    expect(screen.queryByText(/test/i)).not.toBeInTheDocument();
  });
});
