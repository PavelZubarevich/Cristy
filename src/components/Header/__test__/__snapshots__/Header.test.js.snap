// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header top header top is rendered 1`] = `
<body>
  <div>
    <div
      class="header__top"
    >
      <div
        class="container"
      >
        <div
          class="header__topContent"
        >
          <p
            class="header__topLabel"
          >
            FreeShippingOverText
          </p>
          <ul
            class="header__topMenu"
          >
            <li>
              <div
                class="headerDropDown"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="headerDropDown__title "
                  type="button"
                >
                  ru
                </button>
              </div>
            </li>
            <li>
              <div
                class="headerDropDown"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="headerDropDown__title "
                  type="button"
                >
                  USD
                </button>
              </div>
            </li>
            <li>
              <div
                class="headerDropDown"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="headerDropDown__title "
                  type="button"
                >
                  LoginLabel
                </button>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</body>
`;
